nixos/modules/services/misc/navidrome.nix

```diff
@@ -0,0 +1,116 @@
+        example = literalExample ''
+          {
+            LogLevel = "INFO";
+            BaseURL = "/music";
+            ScanInterval = "10s";
+            TranscodingCacheSize = "15MB";
+            MusicFolder = "/Media/Music";
+          };
+        '';
+        description = ''
+        Configuration for Navidrome, see <link xlink:href="https://www.navidrome.org/docs/usage/configuration-options/"/> for supported values.
+      '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "navidrome";
+        description = "User account under which Navidrome runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "navidrome";
+        description = "Group account under which Navidrome runs.";
+      };
+
+      musicFolderPermissions = mkOption {
+        type = types.str;
+        default = "775";
+        description = ''
+        The permissions to set for the music folder.
+        They will be applied on every service start.
+        '';
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.settings.MusicFolder}' '${cfg.musicFolderPermissions}' '${cfg.user}' '${cfg.group}' - -"
+    ];
+
+    environment.etc."navidrome/settings.json".text = builtins.toJSON (cfg.settings // {
+      # These settings are always overwritten by NixOS i.e. they are not settable
+      DataFolder = dataFolder;
+      ConfigFile = configFile;
+    } // (attrsets.optionalAttrs (attrsets.hasAttrByPath ["Port"] cfg.settings) {
+      Port = builtins.toString cfg.settings.Port;
+    }));
+    
+    systemd.services.navidrome = {
+      description = "Navidrome Music Server and Streamer compatible with Subsonic/Airsonic";
+      after = [ "remote-fs.target" "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        ND_CONFIGFILE = configFile;
+      };
+      serviceConfig = {
+        ExecStart = "${pkgs.navidrome}/bin/navidrome";
+        WorkingDirectory = dataFolder;
+        TimeoutStopSec = "20";
+        KillMode = "process";
+        Restart = "on-failure";
+        User = cfg.user;
+        Group = cfg.group;
+        DevicePolicy = "closed";
+        NoNewPrivileges= " yes";
+        PrivateTmp = "yes";
+        PrivateUsers = "yes";
+        ProtectControlGroups = "yes";
+        ProtectKernelModules = "yes";
+        ProtectKernelTunables = "yes";
+        RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6";
+        RestrictNamespaces = "yes";
+        RestrictRealtime = "yes";
+        SystemCallFilter = "~@clock @debug @module @mount @obsolete @privileged @reboot @setuid @swap";
+        ReadWritePaths = dataFolder;
+      };
+    };
+
+    users.users = optionalAttrs (cfg.user == "navidrome") ({
+      navidrome = {
+        description = "Navidrome service user";
+        name = cfg.user;
+        group = cfg.group;
+        home = dataFolder;
+        createHome = true;
```

Comment:
Please use `serviceConfig.StateDirectory = "navidrome";` instead.

Reply:
Could you provide more details? Have I to use that line instead of which one(s)?
I can't find any reference on the manual.

Reply:
I think the nixos manual assumes people are **very** familiar with `systemd` and `systemd` conventions... maybe it shouldn't be.

If you include `systemd.services.navidrome.serviceConfig.StateDirectory = "navidrome";` as part of this module then `systemd` will ensure `/var/lib/navidrome` exists and is owned by `cfg.user:cfg.group` with appropriate permissions **every** time this service starts. Using `users.users.<name>.createHome` will only make sure the folder exists when you run `nixos-rebuild switch` (etc...), so not as good already. Overall it is just "better" to use `StateDirectory` if you want to create `/var/lib/$folder` for a `systemd` service.

Reply:
So I removed the `home` and `createHome` lines and added `StateDirectory = dataFolder` in `systemd.services.navidrome.serviceConfig.StateDirectory`.

Reply:
`StateDirectory` doesn't take absolute paths - it takes folder names relative to `/var/lib`. In your case you should set `StateDirectory = "navidrome";` or better yet `StateDirectory = baseNameOf dataFolder;`.